docs: add note on publishing libraries in ivy

ajitsinghkaler opened the pull request:
Libraries are still built with view engine even after Ivy being the default engine for angular apps stating that explicitly and the reason for it.

Fixes #35625

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Libries built using VE not stated explicitly

Issue Number: N/A


## What is the new behavior?
Libraries built using VE stated explicitly

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ajitsinghkaler commented:
@alan-agius4 please have a look

alan-agius4 commented on aio/content/guide/creating-libraries.md:
```diff
@@ -109,7 +109,7 @@ To learn more, see [Schematics Overview](guide/schematics) and [Schematics for
 ## Publishing your library
 
-Use the Angular CLI and the npm package manager to build and publish your library as an npm package. It is not recommended to publish Ivy libraries to NPM repositories. Before publishing a library to NPM, build it using the `--prod` flag which will use the older compiler and runtime known as View Engine instead of Ivy.
+Use the Angular CLI and the npm package manager to build and publish your library as an npm package. It is not recommended to publish Ivy libraries to NPM repositories. Before publishing a library to NPM, build it using the `--prod` flag which will use the older compiler and runtime known as View Engine instead of Ivy. After version 9 the default compiler for angular apps is Ivy but still the def ...
```
I think adding an alert and changing the wording slightly might make it clearer.

```
Use the Angular CLI and the npm package manager to build and publish your library as an npm package. 

Before publishing a library to NPM, build it using the `--prod` flag which will use the older compiler and runtime known as View Engine instead of Ivy.

<code-example language="bash">
ng build my-lib --prod
cd dist/my-lib
npm publish
</code-example>

If you've never published a package in npm before, you must create a user account. Read more in [Publishing npm Packages](https://docs.npmjs.com/getting-started/publishing-npm-packages).

<div class="alert is-important">
It is not recommended to publish Ivy libraries to NPM repositories because Ivy generated code is not backward compatible with View Engine and the internal instructions is not yet stable which will break consumers using a different Angular version from the one used to build the library.
</div>
```

//cc @kapunahelewong can you please review the proposed wording.

Thanks.

ajitsinghkaler replied:
@kapunahelewong can you please check this

kapunahelewong replied:
Thanks, @ajitsinghkaler! I have to defer to @alan-agius4 on the content. (The language is great, too!)

ajitsinghkaler commented:
@kapunahelewong once you confirm the wording I 'll make the changes

ajitsinghkaler commented:
Closing this in favor of #36643

ajitsinghkaler commented:
@alan-agius4 @kapunahelewong made the changes please have a look

ajitsinghkaler commented:
@alan-agius4 I checked linting I don't know why it is failing on commit message

mary-poppins commented:
You can preview 5407f43 at https://pr36556-5407f43.ngbuilds.io/.
You can preview 889b0fa at https://pr36556-889b0fa.ngbuilds.io/.

alan-agius4 reviewed (approved):
LGTM, thanks for your contribution.

alan-agius4 commented:
@ajitsinghkaler, the lint failure should be fixed by https://github.com/angular/angular/pull/36733

Once merge, kindly rebase and let me know, so we can get this PR merged. Thanks.

kapunahelewong reviewed (approved):
This is very helpful, thanks @ajitsinghkaler!

Reviewed-for: global-docs-approvers

gkalpak commented on aio/content/guide/creating-libraries.md:
```diff
@@ -119,6 +121,10 @@ npm publish
 If you've never published a package in npm before, you must create a user account. Read more in [Publishing npm Packages](https://docs.npmjs.com/getting-started/publishing-npm-packages).
 
+<div class="alert is-important">
+It is not recommended to publish Ivy libraries to NPM repositories because Ivy generated code is not backward compatible with View Engine and the internal instructions is not yet stable which will break consumers using a different Angular version from the one used to build the library.
```
I am not sure what is the best way to incorporate it into this page (regular paragraph? alert? etc. - cc @kapunahelewong), but I think it is important to make it more clear that libraries built with View Engine will still be able to be consumed by Ivy apps. I would say something like:

> For now, it is not recommended to publish Ivy libraries to NPM because Ivy generated code is not backward compatible with View Engine, so apps using View Engine will not be able to consume them. Furthermore, the internal Ivy instructions are not yet stable, which can potentially break consumers using a different Angular version from the one used to build the library.
>
> When a published library is used in an Ivy app, the Angular CLI will automatically convert it to Ivy using a tool known as the Angular compatibility compiler (\`ngcc\`). Thus, by publishing your libraries using the View Engine compiler ensures that they can be transparently consumed by both View Engine and Ivy apps.

ajitsinghkaler replied:
@kapunahelewong comments

kapunahelewong replied:
Yes, I agree with @gkalpak's perspective, @ajitsinghkaler. And an alert would be fitting here, imo.

gkalpak reviewed (approved):
Thx, @ajitsinghkaler!

Can you please update the commit message to more accurately reflect what this is about. (The current one sounds a little scary 😛)
E.g.:

> docs: make clear that libraries should not be published with ivy

ajitsinghkaler committed:
docs: add note on publishing libraries in ivy

Libraries are still build using view engine even after Ivy being the default engine for building angular apps. Added note on why libraries are built using VE and how they will be automatically compiled in Ivy using ngcc making it compatible for both

Fixes #35625

ajitsinghkaler commented:
@alan-agius4 made the changes, rebased it and #36733 has been merged.

aikidave commented on aio/content/guide/creating-libraries.md:
```diff
@@ -109,7 +109,9 @@ To learn more, see [Schematics Overview](guide/schematics) and [Schematics for
-Use the Angular CLI and the npm package manager to build and publish your library as an npm package. It is not recommended to publish Ivy libraries to NPM repositories. Before publishing a library to NPM, build it using the `--prod` flag which will use the older compiler and runtime known as View Engine instead of Ivy.
+Use the Angular CLI and the npm package manager to build and publish your library as an npm package. 
+
+Before publishing a library to NPM, build it using the `--prod` flag which will use the older compiler and runtime known as View Engine instead of Ivy.
```
This is a bit confusing. I use the npm package manager to build/publish an npm package. But I can't publish to NPM? What's the difference between the two? Where else might I publish? Internally to my team?

I would also include the basic instructions, even if they aren't that different from the code snippet below. Here's some alternate wording:

To publish your library, use the `ng build` command.

```
ng build <library-name>
```

This command builds your library as an npm package.

### Distributing libraries to NPM

If you want to distribute your library through [NPM](npmjs.com), use the `--prod` flag.

```
ng build <library-name> --prod
```

<div class="alert is-important">
The <code>--prod</code> flag helps ensure your library is compatible with older versions of Angular.
</div>

For more information about publishing to NPM, see [Publishing npm Packages](https://docs.npmjs.com/getting-started/publishing-npm-packages).

gkalpak replied:
I am quite confused by your question(s) :grin:

> I use the npm package manager to build/publish an npm package.

The npm package manager is used for publishing and managing dependencies. The Angular CLI is used for building.

> But I can't publish to NPM?

What do you mean? You can. Are you saying the current wording implies you can't?

> What's the difference between the two?

Which two???

> Where else might I publish? Internally to my team?

There are many places you can "publich" to. There are private npm registries or even alternative registries, such as GitHub Packages. Or you could just push your package to a git repo or anywhere where it can be retrieved from using a URL.

ajitsinghkaler commented on aio/content/guide/creating-libraries.md:
```diff
@@ -119,6 +121,14 @@ npm publish
 
+<div class="alert is-important">
+
+For now, it is not recommended to publish Ivy libraries to NPM because Ivy generated code is not backward compatible with View Engine, so apps using View Engine will not be able to consume them. Furthermore, the internal Ivy instructions are not yet stable, which can potentially break consumers using a different Angular version from the one used to build the library.
```
> If I am reading this correctly, what we're trying to say is: "Do not publish Ivy libraries to NPM, because Ivy-generated code is not compatible with View Engine." Is that correct?

@aikidave Ivy generated code can only be linked against other Ivy code and is not compatible with VE code. to build the application all VE dependencies from NPM must be converted to Ivy dependencies using `ngcc` compiler. So libraries cannot be published as Ivy code because we do not have a tool that can make Ivy code into VE code but we have ngcc which makes VE code into IVy code.





> If so, the statements here may cause some concern. "Internal Ivy instructions are not yet stable?" Does this mean Angular 9 is not stable?

Ivy instructions are still changing they are not that stable that, they will not change on changing versions but these are stable as in a single version they work perfectly but libraries have to work with various versions of angular as opposed to a single version in codebase. which is the main source of problem.

@gkalpak @alan-agius4 Please rectify if I said anything wrong.

gkalpak replied:
What @ajitsinghkaler said is correct.
@aikidave, to put it another way we are trying to say this:
- There are two "modes" of compiling/building your app: ViewEngine and Ivy.
- When building a library for publishing to NPM, we recommend using ViewEngine mode (for the reasons explained above).

ajitsinghkaler commented:
@gkalpak @alan-agius4 any suggestions

ajitsinghkaler commented:
@aikidave comments

ajitsinghkaler commented:
@aikidave do you still have some doubts

ajitsinghkaler commented:
@aikidave comments

ajitsinghkaler commented:
@aikidave @kapunahelewong how to move this forward

alan-agius4 reviewed (approved):
LGTM

kapunahelewong reviewed (approved):
Thank you so much for this, @ajitsinghkaler 🌸
LGTM!
Reviewed-for: global-docs-approvers

alan-agius4 commented:
@caretaker: @kapunahelewong approved!

alan-agius4 commented:
@ajitsinghkaler, so sorry for the delays on this PR. Thanks a lot for your contribution 😊.

I just marked this to get merged!
